pkg/router/controller/status.go

```diff
@@ -0,0 +1,239 @@
+	position := -1
+	updated := make([]routeapi.RouteIngress, 0, len(route.Status.Ingress))
+	for i := range route.Status.Ingress {
+		existing := &route.Status.Ingress[i]
+		if existing.RouterName != name {
+			updated = append(updated, *existing)
+			continue
+		}
+		if position != -1 {
+			changed = true
+			continue
+		}
+		position = i
+	}
+	switch {
+	case position == -1:
+		position = len(route.Status.Ingress)
+		route.Status.Ingress = append(route.Status.Ingress, routeapi.RouteIngress{
+			RouterName: name,
+			Host:       route.Spec.Host,
+		})
+		changed = true
+	case changed:
+		route.Status.Ingress = updated
+	}
+	ingress := &route.Status.Ingress[position]
+	if ingress.Host != route.Spec.Host {
+		ingress.Host = route.Spec.Host
+		changed = true
+	}
+	return ingress, changed
+}
+
+// setIngressCondition records the condition on the ingress, returning true if the ingress was changed and
+// falso if no modification was made.
+func setIngressCondition(ingress *routeapi.RouteIngress, condition routeapi.RouteIngressCondition) bool {
+	for _, existing := range ingress.Conditions {
+		existing.LastTransitionTime = nil
+		if existing == condition {
+			return false
+		}
+	}
+	now := nowFn()
+	condition.LastTransitionTime = &now
+	ingress.Conditions = []routeapi.RouteIngressCondition{condition}
+	return true
+}
+
+func ingressConditionTouched(ingress *routeapi.RouteIngress) *unversioned.Time {
+	var lastTouch *unversioned.Time
+	for _, condition := range ingress.Conditions {
+		if t := condition.LastTransitionTime; t != nil {
+			switch {
+			case lastTouch == nil:
+				lastTouch = t
+			case t.After(lastTouch.Time):
+				t = lastTouch
+			}
+		}
+	}
+	return lastTouch
+}
+
+// recordIngressConditionFailure updates the matching ingress on the route (or adds a new one) with the specified
+// condition, returning true if the object was modified.
+func recordIngressConditionFailure(route *routeapi.Route, name string, condition routeapi.RouteIngressCondition) (*routeapi.RouteIngress, bool, *unversioned.Time) {
+	for i := range route.Status.Ingress {
+		existing := &route.Status.Ingress[i]
+		if existing.RouterName != name {
+			continue
+		}
+		lastTouch := ingressConditionTouched(existing)
+		return existing, setIngressCondition(existing, condition), lastTouch
+	}
+	route.Status.Ingress = append(route.Status.Ingress, routeapi.RouteIngress{RouterName: name})
+	ingress := &route.Status.Ingress[len(route.Status.Ingress)-1]
+	setIngressCondition(ingress, condition)
+	return ingress, true, nil
+}
+
+// hasIngressBeenTouched returns true if the route appears to have been touched since the last time
+func (a *StatusAdmitter) hasIngressBeenTouched(route *routeapi.Route, lastTouch *unversioned.Time) bool {
+	glog.V(4).Infof("has last touch %v for %s/%s", lastTouch, route.Namespace, route.Name)
+	if lastTouch.IsZero() {
+		return false
+	}
+	old, ok := a.expected.Get(route.UID)
```

Comment:
`expected` is a fixed size cache, so it could return false if the number of routes handled by a router are large (`> 1024`).